code/game/objects/structures/displaycase.dm

```diff
@@ -32,73 +33,85 @@
+	. = ..()
+	if(vname in list(NAMEOF(src, open), NAMEOF(src, showpiece), NAMEOF(src, custom_glass_overlay)))
+		update_icon()
+
+/obj/structure/displaycase/handle_atom_del(atom/A)
+	if(A == electronics)
+		electronics = null
+	if(A == showpiece)
+		showpiece = null
+		update_icon()
+	return ..()
+
 /obj/structure/displaycase/Destroy()
-	if(electronics)
-		QDEL_NULL(electronics)
-	if(showpiece)
-		QDEL_NULL(showpiece)
+	QDEL_NULL(electronics)
+	QDEL_NULL(showpiece)
 	return ..()
 
 /obj/structure/displaycase/examine(mob/user)
 	. = ..()
 	if(alert)
 		. += "<span class='notice'>Hooked up with an anti-theft system.</span>"
 	if(showpiece)
-		. += "<span class='notice'>There's [showpiece] inside.</span>"
+		. += "<span class='notice'>There's \a [showpiece] inside.</span>"
 	if(trophy_message)
 		. += "The plaque reads:\n [trophy_message]"
 
-
 /obj/structure/displaycase/proc/dump()
-	if (showpiece)
-		showpiece.forceMove(loc)
-		showpiece = null
+	if(QDELETED(showpiece))
+		return
+	showpiece.forceMove(drop_location())
+	showpiece = null
+	update_icon()
 
 /obj/structure/displaycase/play_attack_sound(damage_amount, damage_type = BRUTE, damage_flag = 0)
 	switch(damage_type)
 		if(BRUTE)
-			playsound(src.loc, 'sound/effects/glasshit.ogg', 75, TRUE)
+			playsound(src, 'sound/effects/glasshit.ogg', 75, TRUE)
 		if(BURN)
-			playsound(src.loc, 'sound/items/welder.ogg', 100, TRUE)
+			playsound(src, 'sound/items/welder.ogg', 100, TRUE)
 
 /obj/structure/displaycase/deconstruct(disassembled = TRUE)
 	if(!(flags_1 & NODECONSTRUCT_1))
 		dump()
 		if(!disassembled)
-			new /obj/item/shard( src.loc )
+			new /obj/item/shard(drop_location())
 			trigger_alarm()
 	qdel(src)
 
 /obj/structure/displaycase/obj_break(damage_flag)
 	if(!broken && !(flags_1 & NODECONSTRUCT_1))
 		density = FALSE
-		broken = 1
-		new /obj/item/shard( src.loc )
+		broken = TRUE
+		new /obj/item/shard(drop_location())
 		playsound(src, "shatter", 70, TRUE)
 		update_icon()
 		trigger_alarm()
 
+///Anti-theft alarm triggered when broken.
 /obj/structure/displaycase/proc/trigger_alarm()
-	//Activate Anti-theft
-	if(alert)
-		var/area/alarmed = get_area(src)
-		alarmed.burglaralert(src)
-		playsound(src, 'sound/effects/alert.ogg', 50, TRUE)
-
-/obj/structure/displaycase/update_icon()
-	var/icon/I
-	if(open)
-		I = icon('icons/obj/stationobjs.dmi',"glassbox_open")
-	else
-		I = icon('icons/obj/stationobjs.dmi',"glassbox0")
-	if(broken)
-		I = icon('icons/obj/stationobjs.dmi',"glassboxb0")
+	if(!alert)
+		return
+	var/area/alarmed = get_area(src)
+	alarmed.burglaralert(src)
+	playsound(src, 'sound/effects/alert.ogg', 50, TRUE)
+
+/obj/structure/displaycase/update_overlays()
+	. = ..()
 	if(showpiece)
-		var/icon/S = getFlatIcon(showpiece)
-		S.Scale(17,17)
-		I.Blend(S,ICON_UNDERLAY,8,8)
-	src.icon = I
-	return
+		var/mutable_appearance/showpiece_overlay = mutable_appearance(showpiece.icon, showpiece.icon_state)
+		showpiece_overlay.copy_overlays(showpiece)
```

Comment:
Why not full appearance copy ?

Reply:
don't need layer + plane pulled from the object, only the icon + its overlays since we're gonna put another glass sprite over the item anyway (hence the order they're added.)